core/src/bytecode/ast/mod.rs

```diff
@@ -266,6 +266,28 @@ impl AstAlloc {
         }
     }
 
+    /// Allocate an AST element in the arena.
+    ///
+    /// [Self] never guarantees that all destructors are going to be run when using such a generic
+    /// allocation function. We don't want to allocate values that need to be dropped through this
+    /// method, typically because they own heap-allocated data, such as numbers or parse errors.
+    /// That's why we use a marker trait to specify which types can be allocated freely. Types that
+    /// need to be dropped have a dedicated method for allocation.
+    pub fn alloc<T>(&self, value: T) -> &T {
+        self.generic_arena.alloc(value)
+    }
+
+    /// Allocate a sequence of AST elements in the arena.
+    ///
+    /// See [Self::alloc].
+    pub fn alloc_iter<T, I>(&self, iter: I) -> &[T]
+    where
+        I: IntoIterator<Item = T>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        self.generic_arena.alloc_slice_fill_iter(iter)
+    }
+
```

Comment:
Those functions are a bit orthogonal to the matter, but they were introduced in the commit cherry picked here from #2083 and it was easier to keep them. After #2083 is merged I'll make a pass on the `AstAlloc` API to trim it down a bit anyway.

Reply:
That's why I talk about a marker trait that is yet to be written, but will exist in the future.